### ISSUE TITLE
"Too many open files" error, regardless of size parameter

### ISSUE THREAD

**Issue by david-cliqz** [kennethreitz/grequests#54]
I am using grequests to test the performance of my service. Both the service and the machine from which I start the requests are inside Amazon EC2. It seems that I cannot start more requests than `ulimit -n`, even though I explicitly specify the size parameter for `map() / imap()`

My setup is roughly like this:

```
headers = {'Content-type': 'application/json', 'Accept': 'text/plain'}
params = [json.dumps({'url': urls[i]}) for i in urls]
reqs = [grequests.post(SERVICE_URL, data=params[i % len(params)], headers=headers)
            for i in xrange(no_requests)]
res = grequests.imap(reqs, size=200)
for r in res:
    ...
    r.close()
```

Doesn't the `size` parameter here tell `grequests` that it should issue no_requests, 200 at a time?

**Comment by luthur** [comment]
@rtdean fix my problem

**Comment by cyberstip** [comment]
For posterity, just adding a session to the request seems to fix the issue (as @rtdean mentioned):

session = requests.Session()
reqs = (grequests.get(url, session=session) for url in urls)
results = grequests.map(reqs, size=50)

**Comment by bbbco** [comment]
It would be great if you could add this info to a FAQ section in the README

**Comment by skasturi** [comment]
We just faced this issue on our servers bringing down the whole service. It would be good to have it in the README. Thank you!

**Comment by amitsquare** [comment]
Please do mention using Session Object in README .As till you complete system shutdown , user won't be searching the issue & end up here on this issue thread .

**Comment by jontesek** [comment]
@amitsquare @skasturi @bbbco Feel free to create a PR for this :).

**Comment by mattearnshaw** [comment]
the `Session` fix does not work for me, using requests 2.18.1, grequests version 0.30.

**Comment by spyoungtech** [comment]
To provide some further context, this is (or at least was) a problem with HTTPS libraries in general... See requests/requests#239 shazow/urllib3#291 

As far as I can tell, this is not a direct issue with grequests, so I'm going to close the issue. As far as updating the README... Personally, I'm not sure it's worthwhile adding to the currently concise README. I see no particular reason to document it than any other issue not specifically within grequest's domain that affects HTTP libraries in general.  Though, if someone feels this is particularly useful for grequests users, please feel free to submit a PR.
